primitives/npos-elections/solution-type/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+	let SolutionDef {
+		vis,
+		ident,
+		count,
+		voter_pages,
+		voter_type,
+		target_type,
+		weight_type,
+		compact_encoding,
+	} = syn::parse_macro_input!(item as SolutionDef);
+
+	let imports = imports().unwrap_or_else(|e| e.to_compile_error());
+
+	let def = match voter_pages {
+		None => single_page::generate(
+			vis.clone(),
+			ident.clone(),
+			count,
+			voter_type.clone(),
+			target_type.clone(),
+			weight_type.clone(),
+			compact_encoding,
+		)
+		.unwrap_or_else(|e| e.to_compile_error()),
+		Some(p) if p >= 2 => {
+			let single_page_ident = quote::format_ident!("{}Page", ident);
+			let single = single_page::generate(
+				vis.clone(),
+				single_page_ident.clone(),
+				count,
+				voter_type.clone(),
+				target_type.clone(),
+				weight_type.clone(),
+				compact_encoding,
+			)
+			.unwrap_or_else(|e| e.to_compile_error());
+			let multi = multi_page::generate(
+				vis,
+				ident.clone(),
+				single_page_ident,
+				count,
+				p,
+				voter_type.clone(),
+				target_type.clone(),
+				weight_type.clone(),
+			)
+			.unwrap_or_else(|e| e.to_compile_error());
+			quote!(#single #multi)
+		},
+		_ => syn_err("#[pages(_)] can only accept values larger than 2.").to_compile_error(),
+	};
+
+	quote!(
+		#imports
+		#def
+	)
+	.into()
+}
+
+struct SolutionDef {
+	vis: syn::Visibility,
+	ident: syn::Ident,
+	voter_pages: Option<u8>,
+	voter_type: syn::Type,
+	target_type: syn::Type,
+	weight_type: syn::Type,
+	count: usize,
+	compact_encoding: bool,
+}
+
+fn check_attributes(input: ParseStream) -> syn::Result<(bool, Option<u8>)> {
+	let attrs = input.call(syn::Attribute::parse_outer).unwrap_or_default();
+	if attrs.len() > 2 {
+		return Err(syn_err("compact solution can accept only #[compact] and #[pages]"))
+	}
+
+	let has_compact = attrs.iter().any(|attr| {
+		if attr.path.segments.len() == 1 {
+			let segment = attr.path.segments.first().expect("Vec with len 1 can be popped.");
+			if segment.ident == Ident::new("compact", Span::call_site()) {
+				return true
+			}
+		}
+		false
+	});
+
+	let mut pages = None;
+	for attr in attrs.iter() {
+		if attr.path.segments.len() == 1 {
+			let segment = attr.path.segments.first().expect("Vec with len 1 can be popped.");
+			if segment.ident == Ident::new("pages", Span::call_site()) {
+				let tokens = &attr.tokens;
+				if let Ok(parsed) = syn::parse2::<syn::ExprParen>(tokens.clone()) {
+					if let Ok(parsed_pages) = parse_parenthesized_number::<u8>(parsed) {
+						pages = Some(parsed_pages);
+					} else {
+						return Err(syn_err("Failed to parse a number into u8 in `#[pages(_)]`"))
+					}
+				} else {
+					return Err(syn_err("Failed to parse a number into u8 in `#[pages(_)]`"))
+				}
+			}
+		}
+	}
+
+	Ok((has_compact, pages))
+}
+
+impl Parse for SolutionDef {
+	fn parse(input: ParseStream) -> syn::Result<Self> {
+		// optional #[compact] and #[pages(_)]
+		let (compact_encoding, voter_pages) = check_attributes(input)?;
```

Comment:
how are we able to pass type `ParseStream to `check_attribtutes` here without `clone` and still use it again later in this function? afaict it does not impl `copy`

Reply:
ParseStream is a reference https://substrate.dev/rustdocs/latest/syn/parse/type.ParseStream.html
modified without mutability